src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/HeartbeatDefaultPayload.cs

```diff
@@ -3,92 +3,45 @@
     using System;
     using System.Collections.Generic;
     using System.Linq;
-    using System.Reflection;
+    using System.Threading.Tasks;
 
     internal static class HeartbeatDefaultPayload
     {
-       public static readonly string[] DefaultFields =
+        private static readonly IHeartbeatDefaultPayloadProvider[] DefaultPayloadProviders =
         {
-            "runtimeFramework",
-            "baseSdkTargetFramework"
+            new BaseHeartbeatProperties()
         };
 
-        public static void PopulateDefaultPayload(IEnumerable<string> disabledFields, IHeartbeatProvider provider)
+        public static bool IsDefaultKeyword(string keyword)
         {
-            var enabledProperties = RemoveDisabledDefaultFields(disabledFields);
-
-            var payload = new Dictionary<string, HeartbeatPropertyPayload>();
-            foreach (string fieldName in enabledProperties)
+            foreach (var payloadProvider in DefaultPayloadProviders)
             {
-                try
-                {
-                    switch (fieldName)
-                    {
-                        case "runtimeFramework":
-                            provider.AddHeartbeatProperty(fieldName, GetRuntimeFrameworkVer(), true);
-                            break;
-                        case "baseSdkTargetFramework":
-                            provider.AddHeartbeatProperty(fieldName, GetBaseSdkTargetFramework(), true);
-                            break;
-                        default:
-                            provider.AddHeartbeatProperty(fieldName, "UNDEFINED", false);
-                            break;
-                    }
-                }
-                catch (Exception ex)
+                if (payloadProvider.IsKeyword(keyword))
                 {
-                    CoreEventSource.Log.FailedToObtainDefaultHeartbeatProperty(fieldName, ex.ToString());
+                    return true;
                 }
             }
+
+            return false;
         }
 
-        private static List<string> RemoveDisabledDefaultFields(IEnumerable<string> disabledFields)
+        public static async Task<bool> PopulateDefaultPayload(IEnumerable<string> disabledFields, IEnumerable<string> disabledProviders, IHeartbeatProvider provider)
         {
-            List<string> enabledProperties = new List<string>();
+            bool populatedFields = false;
 
-            if (disabledFields == null || disabledFields.Count() <= 0)
+            foreach (var payloadProvider in DefaultPayloadProviders)
             {
-                enabledProperties = DefaultFields.ToList();
-            }
-            else
-            {
-                enabledProperties = new List<string>();
-                foreach (string fieldName in DefaultFields)
+                if (disabledProviders != null && disabledProviders.Contains(payloadProvider.Name, StringComparer.OrdinalIgnoreCase))
```

Comment:
I don't see unit tests testing this feature - disabledProviders is something newly added right? Need to unit test that it is respected by HeartbeatDefaultPayload.

Reply:
You are correct, I’d skipped making these tests after creating the feature! I’ve created a few extra tests.
Thanks for catching this Cijo!